### PR TITLE
[launcher] Add custom token support

### PR DESCRIPTION
**Pull request by jkl73**
Workload can request a custom audience and custom nonce token thru the unix socket.

### CODE REVIEW

File: launcher/container_runner.go [hunk 1]
```diff
@@ -501,6 +486,16 @@ func (r *ContainerRunner) Run(ctx context.Context) error {
 	}
 
 	r.logger.Printf("EnableTestFeatureForImage is set to %v\n", r.launchSpec.Experiments.EnableTestFeatureForImage)
+	// create and start the TEE server behind the experiment
+	if r.launchSpec.Experiments.EnableOnDemandAttestation {
+		r.logger.Println("EnableOnDemandAttestation is enabled, init the server...")
```
Comment (alexmwu) [inline launcher/container_runner.go:491]: EnableOnDemandAttestation is enabled: initializing TEE server.

File: launcher/agent/agent.go [hunk 2]
```diff
@@ -90,6 +91,35 @@ func (a *agent) Attest(ctx context.Context) ([]byte, error) {
...
+		return nil, err
+	}
+
+	resp, err := a.client.VerifyAttestation(ctx, verifier.VerifyAttestationRequest{
+		Challenge:      challenge,
+		GcpCredentials: principalTokens,
+		Attestation:    attestation,
+		CustomAudience: aud,
+		CustomNonce:    nonce,
+	})
```
Comment (alexmwu) [inline launcher/agent/agent.go:116]: This is heavily duplicated with Attest. We should probably have a Attest call AttestCustom with a nil `aud` and `nonce` or pull the common impl into a helper function.

File: launcher/teeserver/tee_server_test.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"path"
+	"testing"
+
+	"github.com/google/go-tpm-tools/launcher/launcherfile"
+)
+
+func TestGetDefaultToken(t *testing.T) {
```
Comment (alexmwu) [inline launcher/teeserver/tee_server_test.go:15]: The attestHandler should take a path to a default attestation token file. That way you don't have to create a file in a well-known location and risk race conditions (then you can remove the change in ci.yml). You can use https://pkg.go.dev/testing#T.TempDir

File: launcher/agent/agent.go [hunk 4]
```diff
@@ -90,6 +91,35 @@ func (a *agent) Attest(ctx context.Context) ([]byte, error) {
 	return resp.ClaimsToken, nil
 }
 
+func (a *agent) AttestCustom(ctx context.Context, aud string, nonce []string) ([]byte, error) {
```
Comment (alexmwu) [inline launcher/agent/agent.go:94]: Oh this should probably take an `Opts` object, since we will definitely add things to this interface.

Comment (JoshuaKrstic) [reply launcher/agent/agent.go:94]: +1

File: launcher/agent/agent.go (hunk 2 above)
Comment (JoshuaKrstic) [reply launcher/agent/agent.go:116]: I was thinking that we just have one method and if we see there are no custom requests (or perhaps we see sts.googleapis as the aud) we can provide the token we have on file.

File: launcher/util.go [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+import (
+	"context"
+	"fmt"
+
+	"google.golang.org/api/impersonate"
+	"google.golang.org/api/option"
+)
+
+// FetchImpersonatedToken return an access token for the impersonetated service account.
```
[the remark this answers is not in the document]
Comment (jkl73) [reply launcher/util.go:11]: Done

File: launcher/teeserver/tee_server.go [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+			return
+		}
+
+		w.WriteHeader(http.StatusOK)
+		w.Write(tok)
+		return
+	}
+
+	w.WriteHeader(http.StatusBadRequest)
+	w.Write([]byte("check the request format"))
```
[the remark this answers is not in the document]
Comment (jkl73) [reply launcher/teeserver/tee_server.go:118]: Done

File: launcher/agent/agent.go (hunk 2 above)
Comment (jkl73) [reply launcher/agent/agent.go:142]: Merging the two attest function here

File: launcher/agent/agent.go (hunk 4 above)
Comment (jkl73) [reply launcher/agent/agent.go:120]: Added

File: launcher/teeserver/tee_server_test.go (hunk 3 above)
Comment (jkl73) [reply launcher/teeserver/tee_server_test.go:15]: good point, done

**Review by alexmwu — APPROVED** [review]
Nice job!

**Pull request merged**